src/ucs/async/thread.c

```diff
@@ -53,25 +55,50 @@ static void ucs_async_thread_hold(ucs_async_thread_t *thread)
 static void ucs_async_thread_put(ucs_async_thread_t *thread)
 {
     if (ucs_atomic_fadd32(&thread->refcnt, -1) == 1) {
-        close(thread->epfd);
+        ucs_event_set_cleanup(thread->event_set);
         ucs_async_pipe_destroy(&thread->wakeup);
         ucs_timerq_cleanup(&thread->timerq);
         ucs_free(thread);
     }
 }
 
+static void ucs_async_thread_ev_handler(void *callback_data, int event,
+                                        void *arg)
+{
+    ucs_async_thread_t *thread = (ucs_async_thread_t *)((void**)arg)[0];
+    int *is_missed             = (int*)((void**)arg)[1];
+    int fd                     = (int)(uintptr_t)callback_data;
+    ucs_status_t status;
+
+    ucs_trace_async("ucs_async_thread_ev_handler(fd=%d, event=%d)",
+                    fd, event);
+
+    if (fd == ucs_async_pipe_rfd(&thread->wakeup)) {
+        ucs_trace_async("progress thread woken up");
+        ucs_async_pipe_drain(&thread->wakeup);
+        return;
+    }
+
+    status = ucs_async_dispatch_handlers(&fd, 1);
+    if (status == UCS_ERR_NO_PROGRESS) {
+         *is_missed = 1;
+    }
+}
+
 static void *ucs_async_thread_func(void *arg)
 {
     ucs_async_thread_t *thread = arg;
-    struct epoll_event events[UCS_ASYNC_EPOLL_MAX_EVENTS];
     ucs_time_t last_time, curr_time, timer_interval, time_spent;
-    int i, nready, is_missed, timeout_ms;
+    int is_missed, timeout_ms;
+    void *cb_arg[2];
     ucs_status_t status;
-    int fd;
+    unsigned read_events;
 
     is_missed  = 0;
     curr_time  = ucs_get_time();
     last_time  = ucs_get_time();
+    cb_arg[0] = thread;
```

Comment:
instead of using an array of void* as arg, pls create a struct type: 1st elem is `ucs_async_thread_t * thread`, 2nd elem is `int* is_missed`

Reply:
@hiroyuki-sato pls apply @yosefe's comment
thank you in advance!

Reply:
Thanks! I completely forgot. fixed.